src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    internal abstract class AbstractMakeFieldReadonlyDiagnosticAnalyzer<TIdentifierNameSyntax, TConstructorDeclarationSyntax>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TIdentifierNameSyntax : SyntaxNode
+        where TConstructorDeclarationSyntax : SyntaxNode
+    {
+        protected AbstractMakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract bool IsWrittenTo(TIdentifierNameSyntax node, SemanticModel model, CancellationToken cancellationToken);
+        protected abstract bool IsMemberOfThisInstance(SyntaxNode node);
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected void AnalyzeType(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferReadonly, context.Node.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node);
```

Comment:
>var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node); [](start = 12, length = 84)

We should write a meta-analyzer that flags this! We should never register a syntax node analyzer that goes back to the declared symbol and then traverses all the declaring references. That is a guaranteed performance issue and there is invariably a better, performant way to write such analysis.